File: LeetCode_611/Main.php
<?php
class Solution {

    /**
     * @param Integer[] $nums
     * @return Integer
     */
    function triangleNumber($nums) {
        sort($nums);
        $n = count($nums);
        $result = 0;

        for ($k = $n - 1; $k >= 2; $k--) {
            $i = 0;
            $j = $k - 1;

            while ($i < $j) {
                if ($nums[$i] + $nums[$j] > $nums[$k]) {
                    $result += $j - $i; // all pairs (i..j-1, j, k) valid
                    $j--;
                } else {
                    $i++;
                }
            }
        }

        return $result;
    }
}

// Test
$sol = new Solution();
$nums = [48,66,61,46,94,75];
echo $sol->triangleNumber($nums); 
?>

File: LeetCode_611/Main.py
class Solution(object):
    def triangleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums.sort()
        n = len(nums)
        result = 0

        def lower_bound(arr, target):
            lo, hi = 0, len(arr)
            while lo < hi:
                mid = lo + (hi - lo) // 2
                if arr[mid] < target:
                    lo = mid + 1
                else:
                    hi = mid
            return lo

        for i in range(n - 1):
            for j in range(i + 1, n):
                if nums[i] == 0 or nums[j] == 0:
                    continue
                s = nums[i] + nums[j]
                lo = lower_bound(nums, s)
                result += lo - j - 1

        return result
